service/history/timerQueueActiveProcessor.go

```diff
@@ -259,6 +265,27 @@ Update_History_Loop:
 				t.logger.Debugf("Activity TimeoutType: %v, scheduledID: %v, startedId: %v. \n",
 					timeoutType, ai.ScheduleID, ai.StartedID)
 
+				if td.Attempt < ai.Attempt && timeoutType != workflow.TimeoutTypeScheduleToClose {
+					continue
+				}
+
+				var retryTask persistence.Task
+				if timeoutType != workflow.TimeoutTypeScheduleToStart {
+					// ScheduleToStart (queue timeout) is not retriable. Instead of retry, user should set larger ScheduleToStart timeout.
+					retryTask = prepareNextRetry(ai, getTimeoutErrorReason(timeoutType))
+				}
+				if retryTask != nil {
+					msBuilder.UpdateActivity(ai)
```

Comment:
I think we should use the existing pattern to have this logic within mutableState.